runtime/runtime/src/lib.rs

```diff
@@ -1687,67 +1688,137 @@ impl Runtime {
+        }
+        for items in grouped_map.values_mut() {
+            items.sort_by_key(|(tx, _)| tx.transaction.nonce());
+        }
 
-            let tx_result = self.process_transaction(
-                state_update,
-                apply_state,
-                signed_transaction,
-                &cost,
-                &mut processing_state.stats,
-            );
-            let (receipt, outcome_with_id) = match tx_result {
-                Ok(outcome) => outcome,
+        // Process each group in parallel using the frozen snapshot.
+        let group_results: Vec<
+            Result<Vec<(SignedTransaction, VerificationResult)>, InvalidTxError>,
+        > = grouped_map
+            .into_par_iter()
+            .map(|(_key, group_list)| {
+                verify_and_charge_transactions(
+                    &apply_state.config,
+                    state_update,
+                    &group_list,
+                    Some(apply_state.block_height),
+                    apply_state.current_protocol_version,
+                    &frozen_snapshot,
+                )
+            })
+            .collect();
+
+        let mut final_map = HashMap::new();
+        for result in group_results {
+            match result {
+                Ok(list) => {
+                    for (tx, vr) in list {
+                        final_map.insert(tx.get_hash(), vr);
+                    }
+                }
                 Err(e) => {
                     Self::handle_invalid_transaction(
                         e,
-                        &tx_hash,
+                        &near_primitives::hash::CryptoHash::default(),
                         processing_state.protocol_version,
-                        "process_transaction error",
+                        "ephemeral group verify error",
                     )?;
-                    continue;
                 }
-            };
-            if receipt.receiver_id() == signed_transaction.transaction.signer_id() {
-                processing_state.local_receipts.push_back(receipt);
-            } else {
-                receipt_sink.forward_or_buffer_receipt(
-                    receipt,
-                    apply_state,
-                    state_update,
-                    processing_state.epoch_info_provider,
-                )?;
             }
-            let compute = outcome_with_id.outcome.compute_usage;
-            let compute = compute.expect("`process_transaction` must populate compute usage");
-            total.add(outcome_with_id.outcome.gas_burnt, compute)?;
-            if !checked_feature!("stable", ComputeCosts, processing_state.protocol_version) {
-                assert_eq!(total.compute, total.gas, "Compute usage must match burnt gas");
+        }
+
+        // Sequential pass in original order: produce receipts.
+        for tx in &original_list {
+            let tx_hash = tx.get_hash();
+            if let Some(vr) = final_map.get(&tx_hash) {
```

Comment:
Similarly, rather than maintaining maps and indexing back, keep a numeric index in the list of sorted transactions and then `.sort_by` the list by that index. Its gonna be faster than constructing a fresh HashMap and easier to follow too.